### PR TITLE
Improve performance of slice_strings for long strings

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Improves on performance for longer strings with `cudf::strings::slice_strings()` API. The `cudf::string_view::substr` was reworked to minimize counting characters and the gather version of `make_strings_children` is used to build the resulting strings column. This version is already optimized for small and large strings.

Additionally, the code was refactored so the common case of `step==1 and start < stop`  can also make use of the gather approach. Common code was also grouped closer together to help navigate the source file better.

The `slice.cpp` benchmark was updated to better measure large strings with comparable slice boundaries. The benchmark showed performance improvement was up to 9x for larger strings with no significant degradation for smaller strings.

Reference #13048 and #12445

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by davidwendt** [comment]
From slice benchmarks
| rows | width | baseline | new | x |
|:---:|:---:| ---:| ---:| ---:|
| 4096  | 32      |  0.117  |  0.132 | 0.9 |
| 4096  | 64      |  0.142  |  0.137 | 1.0 |
| 4096  | 128     |  0.199  |  0.148 | 1.3 |
| 4096  | 256     |  0.330  |  0.176 | 1.9 |
| 4096  | 512     |  0.603  |  0.233 | 2.6 |
| 4096  | 1024    |   1.26  |  0.319 | 3.9 |
| 4096  | 2048    |   2.64  |  0.553 | 4.8 |
| 4096  | 4096    |   5.55  |   1.02 | 5.4 |
| 4096  | 8192    |   12.4  |   2.54 | 4.9 |
| 32768  | 32     |  0.119  |  0.135 | 0.9 |
| 32768  | 64     |  0.147  |  0.141 | 1.0 |
| 32768  | 128    |  0.210  |  0.156 | 1.3 |
| 32768  | 256    |  0.363  |  0.189 | 1.9 |
| 32768  | 512    |  0.669  |  0.271 | 2.5 |
| 32768  | 1024   |   1.50  |  0.390 | 3.8 |
| 32768  | 2048   |   3.56  |  0.850 | 4.2 |
| 32768  | 4096   |   7.46  |   1.63 | 4.6 |
| 32768  | 8192   |   14.8  |   3.14 | 4.7 |
| 262144  | 32    |  0.190  |  0.171 | 1.1 |
| 262144  | 64    |  0.324  |  0.199 | 1.6 |
| 262144  | 128   |   1.38  |  0.288 | 4.8 |
| 262144  | 256   |   7.38  |   1.47 | 5.0 |
| 262144  | 512   |   22.7  |   5.18 | 4.4 |
| 262144  | 1024  |   52.9  |   5.80 | 9.1 |
| 262144  | 2048  |    139  |   18.9 | 7.4 |
| 262144  | 4096  |    293  |   39.2 | 7.5 |
| 2097152  | 32   |  0.780  |  0.479 | 1.6 |
| 2097152  | 64   |   1.72  |  0.597 | 2.9 |
| 2097152  | 128  |   8.54  |   1.10 | 7.8 |
| 2097152  | 256  |   53.2  |   7.34 | 7.2 |
| 2097152  | 512  |    160  |   32.4 | 4.9 |
| 16777216  | 32  |   5.75  |   2.82 | 2.0 |
| 16777216  | 64  |   13.1  |   3.74 | 3.5 |

File: cpp/src/strings/slice.cu [hunk 1]
```diff
@@ -30,12 +30,44 @@
...
+ *
+ * This computes the output size and resolves the substring
+ */
+template <typename IndexIterator>
+struct substring_from_fn {
+  column_device_view const d_column;
+  IndexIterator starts;
+  IndexIterator stops;
+
+  __device__ string_view operator()(size_type idx)
```
Comment (ttnghia) [inline cpp/src/strings/slice.cu:52]: `const`?

File: cpp/src/strings/slice.cu [hunk 2]
```diff
@@ -30,12 +30,44 @@
...
+/**
+ * @brief Function logic for compute_substrings_from_fn API
+ *
+ * This computes the output size and resolves the substring
+ */
+template <typename IndexIterator>
+struct substring_from_fn {
+  column_device_view const d_column;
+  IndexIterator starts;
+  IndexIterator stops;
```
Comment (ttnghia) [inline cpp/src/strings/slice.cu:50]: `const`?

File: cpp/src/strings/slice.cu [hunk 3]
```diff
@@ -92,14 +124,50 @@ struct substring_fn {
     char* d_buffer  = d_chars ? d_chars + d_offsets[idx] : nullptr;
     auto itr        = begin;
     while (step > 0 ? itr < end : end < itr) {
-      bytes += bytes_in_char_utf8(*itr);
-      if (d_buffer) d_buffer += from_char_utf8(*itr, d_buffer);
+      if (d_buffer) {
+        d_buffer += from_char_utf8(*itr, d_buffer);
+      } else {
+        bytes += bytes_in_char_utf8(*itr);
+      }
```
Comment (ttnghia) [inline cpp/src/strings/slice.cu:131]: Was this a bug before? Since `bytes` is computed in both situations?

Comment (davidwendt) [reply cpp/src/strings/slice.cu:131]: No. The `bytes` value is not actually used in the second path so I am just saving a few instructions.

File: cpp/src/strings/slice.cu [hunk 4]
```diff
@@ -92,14 +124,50 @@ struct substring_fn {
...
+ * This can also be called with constant value iterators to handle special
+ * slice functions if possible.
+ *
+ * @tparam IndexIterator Iterator type for character position values
+ *
+ * @param d_column Input strings column to substring.
+ * @param starts Start positions index iterator.
+ * @param stops Stop positions index iterator.
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @param mr Device memory resource used to allocate the returned column's device memory.
```
Comment (ttnghia) [inline cpp/src/strings/slice.cu:153]: I assume that we don't want period for these tags.

**Review by ttnghia — APPROVED** [review]

**Review by elstehle — APPROVED** [review]

**Pull request merged**